src/porepy/numerics/time_step_control.py

```diff
@@ -404,6 +404,15 @@ def __repr__(self) -> str:
 
         return s
 
+    def final_time_reached(self) -> bool:
+        """Check whether the time manager has reached the end of the schedule.
+
+        Returns:
+            Whether the final time has reached or been overstepped.
+
+        """
+        return self.time > self.time_final or np.isclose(self.time, self.time_final)
```

Comment:
This will become a problem if the time step is tiny compared to the final time. Will it work if we insist on using a relative tolerance (so, set `np.isclose(..., atol=self.time_final * 1e-30)` or something), and then scale the relative tolerance with the time step, `np.isclose(..., rtol=self.dt * 1e-8)`. This will also fail if the time step must be severely cut just before the end of simulation (thus `dt` is dramatically decreased), but it should be more robust.

Let me know what you think about this, but don't accept my logic by default, I don't trust it myself.

Reply:
The scale you mention is already performed in the implementation of [np.isclose](https://numpy.org/doc/stable/reference/generated/numpy.isclose.html). Tolerances for float point equality were exposed to the user with default:

- rtol: float = 1e-10
- atol: float = 1e-16

The parametrization of the test was updated to cover time steps and final times values of the order or (1e-14,1e0,1e+14). See comment above.

Reply:
If `atol=self.time_final * 1e-30` should make sense, why not setting to zero and entirely deactivating the absolute criterion?

I think I like @keileg 's suggestion more regarding relative tolerances. Just using a fixed value for `rtol` without relaing to the time step size will not necessarily fix the issue when the final time is large while the time step size is small (although in a practical situation one most likely would be absolutely fine if the final simulation time is not 10 years, but 10 years minus 0.03 seconds).

Reply:
See comment above.

Reply:
As @jwboth writes, any fixed `rtol` will be insufficient if the problem is sufficiently difficult, or the user has prescribed sufficiently small time steps. Whether it makes any practical difference if the simulation stops `epsilon` time before what was intended is a different matter. 

Come to think of it, precisely the same problem will arise when [adapting the time step to the schedule](https://github.com/pmgbergen/porepy/blob/83581f3a5fdd0bf3358a0c3c2d3ad306edd5b245/src/porepy/numerics/time_step_control.py#L663). Should we fix this at the same time?

Reply:
Update: Adaptation to time step will not be a problem, the implementation is slightly different from what I recalled.